### PR TITLE
Align datetimeindex slicing behaviour with Pandas 2.x

### PR DESCRIPTION
**Pull request by shwina**
Closes https://github.com/rapidsai/cudf/issues/14863.

### CODE REVIEW

File: python/cudf/cudf/tests/test_indexing.py [hunk 1]
```diff
@@ -1299,13 +1299,32 @@ def test_loc_datetime_index(sli, is_dataframe):
...
     got = gd_data.loc[sli]
-
     assert_eq(expect, got)
 
 
+@pytest.mark.parametrize(
+    "sli",
+    [
+        slice("2001", "2020"),
+        slice(None, "2020"),
```
Comment (bdice) [inline python/cudf/cudf/tests/test_indexing.py:1311]: I didn't realize you could pass strings (or non-integer values) to `slice`... I guess this is handled by the `.loc[]`, but it feels funny to me.

File: python/cudf/cudf/core/indexed_frame.py [hunk 2]
```diff
@@ -198,9 +198,18 @@ def _get_label_range_or_mask(index, start, stop, step):
         if start is not None and stop is not None:
             if start > stop:
```
Comment (mroeschke) [inline python/cudf/cudf/core/indexed_frame.py:199]: I think this condition also needs to check that start and stop are in the index

```python
In [2]: df = pd.DataFrame([1, 2, 3], index=pd.DatetimeIndex(["2001", "2009", "2004"]))

In [4]: df.loc["2009":"2001"]
KeyError: 'Value based partial slicing on non-monotonic DatetimeIndexes with non-existing keys is not allowed.'
```

**Review by bdice — APPROVED** [review]
Code looks fine. A test looks funny to me, but I think it's testing an intended behavior that I just haven't seen before.

**Review by galipremsagar — APPROVED** [review]

**Comment by galipremsagar** [comment]
For now, I'm going to merge this PR, but also agree with @mroeschke's review here: https://github.com/rapidsai/cudf/pull/14887#discussion_r1466822135, I think this is okay to be addressed as a follow-up PR.

**Pull request merged**